device/devicegate/filter.go

```diff
@@ -0,0 +1,263 @@
+	copy := make(map[string][]interface{})
+
+	f.lock.RLock()
+	for k, v := range f.filters {
+		var filters []interface{}
+		for filter := range v {
+			filters = append(filters, filter)
+		}
+
+		copy[k] = filters
+	}
+	f.lock.RUnlock()
+
+	return copy
+}
+
+func (f *FilterGate) FiltersCopy() map[string]map[interface{}]struct{} {
+	filtersCopy := make(map[string]map[interface{}]struct{})
+
+	f.lock.RLock()
+	for key, val := range f.filters {
+		valCopy := make(map[interface{}]struct{})
+
+		for k, v := range val {
+			valCopy[k] = v
+		}
+
+		filtersCopy[key] = valCopy
+	}
+	f.lock.RUnlock()
+
+	return filtersCopy
+}
+
+func (f *FilterGate) Filters() map[string]map[interface{}]struct{} {
+	return f.filters
+}
+
+//allows for adding and removing filters
+//Will completely overwrite current filters in place
+func (f *FilterGate) EditFilters(key string, values []interface{}, add bool) bool {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	if add {
+		valuesMap := make(map[interface{}]struct{})
+
+		for _, v := range values {
+			valuesMap[v] = emptyVal
+		}
+
+		f.filters[key] = valuesMap
+
+		return true
+	} else {
+
+		_, ok := f.filters[key]
+
+		if ok {
+			delete(f.filters, key)
+			return true
+		}
+
+		return false
+	}
+}
+
+func (f *FilterGate) AllowConnection(d device.Interface) (bool, device.MatchResult) {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	for filterKey, filterValues := range f.filters {
+
+		// check if filter is in claims
+		if claimsMatch(filterKey, filterValues, d.Metadata()) {
+			return false, device.MatchResult{
+				Location: metadataMapLocation,
+				Key:      filterKey,
+			}
+		}
+
+		// check if filter is in metadata map
+		if metadataMapMatch(filterKey, filterValues, d.Metadata()) {
+			return false, device.MatchResult{
+				Location: claimsLocation,
+				Key:      filterKey,
+			}
+		}
+
+	}
+
+	return true, device.MatchResult{}
+}
+
+func metadataMapMatch(keyToCheck string, filterValues map[interface{}]struct{}, m *device.Metadata) bool {
+	metadataVal := m.Load(keyToCheck)
+	if metadataVal != nil {
+		switch t := metadataVal.(type) {
+		case interface{}:
+			return filterMatch(filterValues, t)
+		case []interface{}:
+			return filterMatch(filterValues, t...)
+
+		}
+	}
+
+	return false
+
+}
+
+func claimsMatch(keyToCheck string, filterValues map[interface{}]struct{}, m *device.Metadata) bool {
+	claimsMap := m.Claims()
+
+	claimsVal, found := claimsMap[keyToCheck]
+
+	if found {
+		switch t := claimsVal.(type) {
+		case interface{}:
+			return filterMatch(filterValues, t)
+		case []interface{}:
+			return filterMatch(filterValues, t...)
+		}
+	}
+
+	return false
+}
+
+func filterMatch(filterValues map[interface{}]struct{}, paramsToCheck ...interface{}) bool {
+	for _, param := range paramsToCheck {
+		_, found := filterValues[param]
+
+		if found {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func (f *FilterGate) ServeHTTP(response http.ResponseWriter, request *http.Request) {
```

Comment:
A better approach is to use black box reuse or composition.  Implement an `http.Handler` that contains a `*FilterGate`:

```
type FilterHandler struct {
    Gate Gate // use an interface is preferred
}

func (fh FilterHandler) ServeHTTP(...) {
    ...
}
```

That way is easier to unit test and maintain over time, since adding features to the `Gate` won't necessarily affect the `http.Handler` code.